src/entities/tile.rs

```diff
@@ -188,22 +129,49 @@ impl LevelGrid {
         &self.grid
     }
 
-    // we cannot store and use the Grid we deserialized, because it may have changed and we don't want to have two representations of the the same Grid
-    pub fn generate_tile_grid_copy<T: GenericReadStorage<Component = Tile>>(
+    pub fn determine_sprite_for<'a, T: GenericReadStorage<Component = Tile>>(
         &self,
-        tile_storage: &T,
-    ) -> Grid {
-        let mut grid = self.grid.clone();
-        Grid {
-            grid: grid
-                .iter_mut()
-                .map(|vec| {
-                    vec.iter_mut()
-                        .map(|entity| (*tile_storage.get(*entity).unwrap()).clone())
-                        .collect()
-                })
-                .collect(),
+        x: i32,
+        y: i32,
+        dictionary: &'a Vec<([[Tile; 3]; 3], String)>,
+        storage: &T,
+    ) -> (&'a str, i32) {
+        let tile = self.get_tile(x as i32, y as i32, storage).unwrap();
+        if let Tile::Ground { concealed: true } = tile {
+            return (CONCEALED, 0);
+        };
+        let mut key = [[Tile::default(); 3]; 3];
+        for delta_x in 0..3 {
+            for delta_y in 0..3 {
+                if let Some(t) =
+                    self.get_tile(x + delta_x as i32 - 1, y + delta_y as i32 - 1, storage)
+                {
+                    key[delta_x][delta_y] = *t;
+                    // if we get a None (aka out of bounds) we want a Tile::Default at this position.
+                    // nothing to do here because of the initialize with Tile::Defaults...
+                }
+            }
+        }
+
+        for rotation in 0..3 {
+            if let Some(result) = util::find_in_vec(&key, &dictionary) {
+                return (result.as_str(), 90 * (rotation + 1));
+            };
+            key = util::rotate_3x3(&key);
         }
+        panic!("Cannot determine sprite for: {:?}", util::rotate_3x3(&key));
+    }
+
+    pub fn get_tile<'a, T: GenericReadStorage<Component = Tile>>(
+        &self,
+        x: i32,
+        y: i32,
+        storage: &'a T,
+    ) -> Option<&'a Tile> {

```

Comment:
you sure you need to specify `'a`?

Reply:
Yes, because the return is a reference to a storage value and can only live as long as the storage lives

Reply:
Yes, because the return is a reference to a storage value and can only live as long as the storage lives